### PR TITLE
esp8266/modpybhspi: Add a HSPI module for hardware SPI support (esp-open-rtos version)

### PR DESCRIPTION
**Pull request by deshipu**
This is a work in progress for adding support for fast hardware-based SPI to the ESP8266 port.

### CODE REVIEW

**Comment by deshipu** [comment]
This is my first micropython module, so any feedback is much appreciated.

**Comment by mladenbruck** [comment]
I'll gladly help to test it with my mcp23s17 board, as soon as I figure out how to upload mycropython firmware to nodemcu  :)

File: esp8266/modpybhspi.c [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        self->baudrate = args[ARG_baudrate].u_int;
+    }
+    if (args[ARG_polarity].u_int != -1) {
+        self->polarity = args[ARG_polarity].u_int;
+    }
+    if (args[ARG_phase].u_int != -1) {
+        self->phase = args[ARG_phase].u_int;
+    }
+
+    spi_mode(HSPI, self->phase, self->polarity);
```
Comment (deshipu) [inline esp8266/modpybhspi.c]: For some reason this conflicts with setting a custom baud rate different than 80Mhz. The clock signal becomes all weird.

**Comment by mcauser** [comment]
Based on: https://github.com/MetalPhreak/ESP8266_SPI_Driver

**Comment by mcauser** [comment]
@mladenbruck here's how I got it working on a WeMos D1 mini, which is similar to the NodeMCU dev board: http://forum.wemos.cc/topic/161/micropython-on-the-esp8266-it-works

**Comment by pfalcon** [comment]
Mu personal preference would be that https://github.com/SuperHouse/esp-open-rtos should be a default source of hardware drivers not available in uPy. It's more clean-room implementation with more consistent code style, and of course more general in coverage of esp8266 hardware.

**Comment by deshipu** [comment]
OK, I will try to replace the library this evening and see how that works. The code indeed looks much cleaner.

**Comment by deshipu** [comment]
One question though -- is the copying of the files into the micropython's tree (license permitting, of course) fine, or should it be a dependency or a git subrepo, or some other way?

**Comment by pfalcon** [comment]
> OK, I will try to replace the library this evening and see how that works.

Thanks for considering this option.

> One question though -- is the copying of the files into the micropython's tree (license permitting, of course) fine, or should it be a dependency or a git subrepo, or some other way?

I'd say, copy. esp-open-rtos is MIT/BSD (and people behind it care about licensing, so it should be clean). Their file headers (e.g. https://github.com/SuperHouse/esp-open-rtos/blob/02c35d8a7153a9f8be895f6a170bccdc1307f0a0/core/esp_spi.c) should have enough information regarding attribution and licensing, and majority of their content won't be of use to us, so to carry it around as a submodule doesn't seem like a good idea. We may also need to patch their code anyway, so a submodule won't help match. Anyway, that's basic idea - I'd start with copying, and if actually use more and more of their content over time, later consider submodule option (based on amount of patching we actually apply).

**Comment by deshipu** [comment]
So I copied all the needed files (10 header files and 2 C files) into the tree (not as clean-room after all), and reworked the glue code to use it. However, now I get:

```
LINK build/firmware.elf
xtensa-lx106-elf-ld: build/firmware.elf section `.iram1.text' will not fit in region `iram1_0_seg'
xtensa-lx106-elf-ld: region `iram1_0_seg' overflowed by 40 bytes
Makefile:185: recipe for target 'build/firmware.elf' failed
make: *** [build/firmware.elf] Error 1
```

when trying to compile it.

**Comment by deshipu** [comment]
Updated the pull request (but it doesn't compile). Any ideas?

**Comment by pfalcon** [comment]
> region `iram1_0_seg' overflowed by 40 bytes

iRAM is scarce resource. You would need to disable something else. E.g., try to remove espapa102.c from compilation (with all deps). We actually need to make all esp8266 modules configurable, so tehy can be enabled/disabled via mpconfigport.h

**Comment by deshipu** [comment]
I tired to exclude the software spi from compilation, but that didn't change anything -- either I didn't remove it from all places, or it doesn't use any iRAM?

**Comment by pfalcon** [comment]
Yes, you need to disable stuff that uses iram. Generally, it's esp*.c .

**Comment by deshipu** [comment]
Wouldn't it be better to fall back to the library I used initially, which seems to not use iRAM either?

**Comment by pfalcon** [comment]
Well, the likely issue you experience is that you didn't tell linker to put new code into flashrom, so it goes into iRAM (see esp8266.ld). Would need to compare libs in detail to tell which one is better (or why MetalPhreak's uses less space). Note that to get fast and predictable timing, code should be in iRAM.

**Comment by deshipu** [comment]
I see, I will look into that. Meanwhile I disabled the espapa102 and got it to compile, trying to figure out why it sends no data and clock signals...

**Comment by deshipu** [comment]
I solved the iram problem by adding the two .c files to the linker configuration -- thanks for the suggestion. I also don't think they need to be fast or have predictable timing -- they call the hardware to do all of the timing-critical work.

However, I can't get the rtos code to work. Either it requires some additional setup that rtos does, or there is something I missed completely there. No matter what settings I use, there is no output on MOSI or CLK, and sometimes random things happen, like the LED on the GPIO0 coming on, or the board resetting. I spent the last two hours on that and I'm starting to doubt my sanity.

**Comment by pfalcon** [comment]
> they call the hardware to do all of the timing-critical work.

Well, SPI itself isn't time-critical even for bit-banging, but wasn't you surprised by the long delays between bytes/blocks? Using iRAM may help (or not) with that.

> However, I can't get the rtos code to work.

Well, there's some bad luck here. It was the same situation with PWM - @dpgeorge looked into taking one from esp-open-rtos, but ended up taking from nodemcu. I guess it's not worth spending more time on that, but I'd defer further comments on that to him. If you can, please post separate PRs for both libraries, so they can be compared. Thanks!

**Comment by deshipu** [comment]
I could also lift this code from nodemcu -- it uses HSPI by default, and in fact has no software SPI, I think. At first sight the code looks good. Do you think it makes sense to make a third version of this?

**Comment by pfalcon** [comment]
> Do you think it makes sense to make a third version of this?

Nope, I personally would find that daunting experience ;-). Time would better be spent on testing it (e.g. with various HW devices, speeds, etc.) or working on something else. That's IMHO of course.

**Comment by deshipu** [comment]
Well, I have a small problem with that. While I did test output both with several LCD and OLED displays, and with a logic analyzer, I don't think I have any SPI device that I could read from. I will try with PyBoard set to slave, but that's not really representative of the smaller devices out there.

**Comment by pfalcon** [comment]
Don't you have nRF24L01 around?

**Comment by deshipu** [comment]
Great idea. I don't have nRF24L01, but I have a bunch of cheap "pin-compatible" modules. The thought of connecting a radio module to a wifi board somehow never occurred to me.

**Comment by deshipu** [comment]
Looking at the NodeMCU's implementation, I can see two big advantages: it's only 3 files (2 headers and one C) containing only what we need and no extra stuff, and it also has an SPI slave implementation.

**Comment by torwag** [comment]
@deshipu:
As for your SPI work, you might be interested in getting a buspirate board from Dangerous prototypes. http://dangerousprototypes.com/docs/Bus_Pirate
It allows you to read and write and sniff a SPI bus (and many other chip to chip buses)  on a rather low level. You can set speed, protocol deviations, even things like pull ups and other low level stuff. Perfect for debugging,  bug-tracking and head scratching sessions. I am not associated in any way, thus this is no ad ;) Actually, one idea which circulates in my mind for some time, create something like the buspirate with the power of micropython ;)

**Comment by deshipu** [comment]
@torwag thanks for the hint. I considered getting that, but in the end I decided on a logic analyzer, which lets me look at the exact signals, and the software has SPI/I2C/etc. decoding.

**Comment by deshipu** [comment]
The other version won.

**Pull request closed without merging**